[palvo.v 최환] 지하철 서비스 코드리뷰 부탁드립니다.

choihwan2 opened the pull request:
안녕하세요 파블로입니다!
이번 미션을 진행하면서 이해가 잘 가지 않고 궁금했던 부분은 인증에 관련된 부분이였던 것 같습니다.
페어와 함께 얘기를 하다보니어떤 노선의 구간(line_id)인지 구간객체(Section) 만으로 확인 할 수 있도록 기존의 뼈대코드에서 Secition 객체에 line_id 를 추가하여 코드를 조금 수정해봤습니다.
질문은 코드에 적어놓도록 하겠습니다!!

kakao-yell-ow committed:
feat : 1단계 경로 조회 기능 구현

- PathService, Graph, Path 생성
- Lines 일급 컬렉션 생성
- Section의 간선 역할을 위해 DefaultWeightedEdge 상속
- PathController 최단 경로를 리턴하는 findPath 메서드 추가

kakao-yell-ow committed:
feat : 2단계 회원과 인증 관련 기능 구현

- README.md 파일에 요구사항 추가
- AuthService 인증 기능 구현
- AuthController login 메서드 구현
- AuthenticationPrincipalArgumentResolver resolveArgument 구현
- MemberService findByEmail 메서드 추가
- MemberDao CRUD 예외 상황 처리 로직 구현
- MemberController Error Handling 기능 추가

kakao-pavlo-v committed:
feat : 즐겨찾기 기능 구현

- FavoriteService, FavoriteDao, Favorite 추가
- FavoriteResponse 생성자 추가
- FavoriteController 에서 CRD API 구현
- 즐겨찾기 생성됨 검증 테스트 코드 추가

kakao-pavlo-v committed:
feat : 3단계 거리에 따른 요금 계산 기능 구현

- Section에 LineId 멤버변수 추가
- Section 변경에 따른 LineService, LineDao, SectionDao, Section, Sections 수정
- LineResponse getExtraFare 메서드 추가
- LoginMember 연령에 따른 할인율 계산 기능 추가
- PathService 로그인 비로그인에 따른 요금 계산 기능 구현
- Graph 구간에 포함된 노선 조회 기능 추가
- Path 요금 계산 기능 구현
- PathResponse 정적 팩토리 메소드 오버로딩
- PathController 로그인에 해당하는 경로조회 API 구현
- FareUtil 거리에 해당하는 요금 계산하는 클래스 구현

- DataLoader 노선 생성자 형식 변경
- 테스트 코드 수정

kakao-yell-ow committed:
refactor : 메시지 및 정수 상수화 진행

kakao-yell-ow committed:
refactor : CustomException 생성 및 GlobalControllerAdvice 로 Error Handling 하도록 수정

kakao-pavlo-v committed:
refactor : 코드 리팩토링

- final 키워드 빠진부분에 추가
- 상수 변수화

kakao-yell-ow committed:
refactor : PathController GetMapping 로직변경 및 PathService null 처리 로직 추가

kakao-pavlo-v committed:
refactor : Merge Conflict

choihwan2 commented on src/main/java/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -24,6 +28,15 @@ public boolean supportsParameter(MethodParameter parameter) {
+        String accessToken = AuthorizationExtractor.extract(httpServletRequest);
+        if(accessToken == null || accessToken.equals("null")) {
+            return null;
+        }
```
이 if문의 사용이 가장 큰 궁금점이였던 것 같습니다. 프론트 코드가 제공되다 보니 어떻게 작성되어있는지 모르겠지만 노선관리 확인 부분에서 Token의 값이 `Authorization` 헤더값이 없더라도 "null" 이라는 문자열이 반환되는 것이 확인 되었는데, 이게 프론트의 문제인지 아니면 제가 코드를 작성하면서 놓친 부분이 있어서 이렇게 처리를 해야하는건지 정확하게는 모르겠지만, 제 개인적인 생각으로는 프론트에서 로그인을 한 상태에서는 'Authorization'의 헤더를 언제나 보내줘야하는 것이 아닌

1. 로그인한 경우 ('Authorization' 헤더를 같이 requset에 추가해서 보내줌)
2. 로그인 하지 않은 상태('Authorization' 헤더를 추가하지 않음.

이렇게 두개의 리퀘스트가 다르게 요청되어야 하는 것이 아닌가요?? 일단은 예외를 막기위해서 이렇게 코드를 작성해놓았는데, 이런 상황이 현업에서 발생하는 경우도 있는지, 그럴 경우 백엔드단에서만 해결할 수 있는 방법이 있는지 궁금합니다!

kriery replied:
결과부터 말씀드리자면, 필요합니다 (우울하게도..). 이 if문은 FE코드의 방어로직이라고 봐야 하는데요.
현업도 사람이 하는 일이라서 늘 **완전무결**하지만은 않습니다 😭 
원인은, frontend/src/api/index.js 에 있어요
```javascript
        Authorization: `Bearer ${localStorage.getItem('token')}` || '',
```
(이런게 4개) 아마 다음과 같이 되어야 할 것 같은데요.
```javascript
        Authorization: localStorage.getItem('token') && `Bearer ${localStorage.getItem('token')}` || '',
```
아무튼 파블로의 생각처럼 되어야 하는게 맞긴 하지만, 
**현업**에서는 한 부서에서 FE/BE를 모두 같이 하는 경우는 많지가 않고 협업이 필요한 일인데,
제보하는데로 바로바로 수정이 되면 참 좋겠지만, 이게 유기적으로 빠르게 해결되지 않을 때가 많고 서비스는 문제없이 되어야 하다 보니까.
불가피하게 **방어로직**이라는 개념이 들어가기도 합니다.
(근데 사실 저라면 그냥 `if`문 빼고 아래로 내려서 `checkToken`에서 `Exception`이 날아가게 둘거 같네요.. FE에서 고쳐주겠지..)

choihwan2 replied:
좋은 답변 너무 감사합니다☺️
저와 페어도 사실 `if`문을 빼고 `checkToken` 에서 예외가 날아가게 두고싶었는데... 그렇게 하면 만들어둔 테스트코드가 통과가 되지않아서😭.. 방어로직이라는 개념에 대해서 배운것 같아서 기쁘네요! 감사합니다!

choihwan2 commented on src/main/java/subway/member/dao/MemberDao.java:
```diff
@@ -39,12 +45,18 @@ public Member insert(Member member) {
+        if(updatedRow != 1) {
+            throw new NotExistsDataException(NOT_EXISTS_USER_INFO_ERROR_MESSAGE);
+        }
     }
```
페어와 조금 의견이 갈렸던 부분이 였는데, 혹시 dao 객체에서 예외를 처리하지 않고 Service 객체로 예외 상황을 넘겨준후 Service에서 처리하는지, 아니면 지금처럼 dao에서 예외를 처리하는지 궁금합니다.

kriery replied:
(try~catch 와 비교해서)
`throw Exception`은 예외를 **처리** 한다기 보다는, 예외를 **유발** 시켰다고 보는 편이 맞을것 같긴해요.
지금처럼 `@ControllerAdvice`가 있어서 (또는 다른 형태로) `@ExceptionHandler`가 동작하고 있는 걸로 봐서는,
예외의 **처리** 주체가 아니라 예외의 **유발** 주체를 질문하신것 같네요.
다소 주관적인 질문이예요 (그래서 페어랑 의견이 엇갈렸을것 같아요).
이는 꼭 DAO에서 해야 한다, Service에서 해야 한다는 규칙이나 관례가 있기 보단, "이 상황을 **예외**로 판단할 것인가"와 관련이 있습니다.
- ex-1: 마지막 로그인을 한지 1년이 넘은 사용자를 휴면처리하는 DAO method
-> update 콜의 결과는 DML에 영향받은 row 수인데, 1년이 넘은 사용자가 없을 수도 있고 많은 수도 있고, 없더라도 문제가 되지 않으니까 그냥 update 된 결과 갯수만 반환해도 무방할테죠
- ex-2: (지금처럼) member 정보를 변경함
-> 내 정보를 변경하는 DML은 항상 1건만 영향을 받아야 합니다. 만약 2건을 update했거나, 0건이면 예외를 **유발**하여 해당 Transaction을 빠르게 되돌릴 (roll-back) 필요가 있어요.

choihwan2 replied:
용어 자체(처리와 유발) 을 잘못 알고 사용하고 있었네요. 어디서 유발을 해야한다는 규칙보다는 판단에 초점을 맞추라는 조언 감사합니다!!

kriery commented on src/main/java/subway/member/ui/MemberController.java:
```diff
@@ -1,60 +1,63 @@
 import java.net.URI;
 
 @RestController
+@RequestMapping("/members")
```
깔끔하게 정리했네요 👍

kriery commented on src/main/java/subway/path/dto/PathResponse.java:
```diff
@@ -18,6 +20,13 @@ public PathResponse(List<StationResponse> stations, int distance, int fare) {
+    }
+    public static PathResponse of(Path path, LoginMember loginMember) {
+        return new PathResponse(path.getStations(), path.getDistance(), path.getTotalFare(loginMember));
+    }
```
`StaticFactoryMethod` 💯

kriery commented on src/main/java/subway/line/domain/Section.java:
```diff
@@ -12,10 +15,8 @@ public Section() {
         this.id = id;
-        this.upStation = upStation;
-        this.downStation = downStation;
-        this.distance = distance;
```
많은 일을 하는 생성자 (Long id, Station upStation, Station downStation, int distance) 에서
적은 일을 하는 생성자 (Station upStation, Station downStation, int distance) 를 이용하기 보다는,
그 반대로 이용하는 편이 더 이득이 많아요 (property를 final로 바꿀 수도 있죠)

choihwan2 replied:
감사합니다! 변경해보았어요!

kriery commented on src/main/java/subway/line/dao/LineDao.java:
```diff
@@ -53,13 +55,13 @@ public Line findById(Long id) {
-        String sql = "update LINE set name = ?, color = ? where id = ?";
-        jdbcTemplate.update(sql, new Object[]{newLine.getName(), newLine.getColor(), newLine.getId()});
+        String sql = "update LINE set name = ?, color = ?, extra_fare = ? where id = ?";
+        jdbcTemplate.update(sql, new Object[]{newLine.getName(), newLine.getColor(),newLine.getFare(), newLine.getId()});
```
`jdbcTemplate.update(String sql, @Nullable Object... args)` 형태라서,
new Object[] 없이 가변 인자 형태로 호출 가능합니다 😄 
```suggestion
        jdbcTemplate.update(sql, newLine.getName(), newLine.getColor(),newLine.getFare(), newLine.getId());
```

kriery commented on src/main/java/subway/line/dao/LineDao.java:
```diff
@@ -69,14 +71,14 @@ public void update(Line newLine) {
     }
 
     private Line mapLine(List<Map<String, Object>> result) {
         if (result.size() == 0) {
-            throw new RuntimeException();
+            throw new NotExistsDataException("해당 노선에 해당하는 구간이 존재하지 않습니다.");
```
exception customizing 💯

kriery commented on src/main/java/subway/path/domain/Path.java:
```diff
@@ -0,0 +1,41 @@
+public class Path {
+
+    public static final int DEDUCTIONS = 350;
+    private final List<StationResponse> stations;
```
domain 객체인 `Path`를 구성하는 요소에 `StationResponse`가 있다는건 맞지 않은것 같아요.
`Station`을 이용하는 편이 더 명확하겠습니다.

choihwan2 replied:
그렇네요!! 수정하겠습니다!

kriery reviewed (changes requested):
안녕하세요 파블로. 코드 진행이 깔끔하고 어색한 부분이 없어서 따라가기 편했네요.
이젠 따로 얘기 안해도 일급 컬렉션을 자연스럽게 이용하시는것 같습니다.
`path.domain.Graph` / `path.domain.Path` 두가지 도메인 객체는 로직을 담고 있는게 있는데
요 부분에 대한 테스트 코드를 보강해주셨으면 합니다.
(다음 단계에 반영해주셨으면 좋겠다는 답변을 드리고 싶으나, 여기가 마지막이라서 그럴 수가 없네요)
금방 반영해주실거니까 곧 봐요!

kakao-pavlo-v committed:
refactor : 코드 리뷰바탕으로 수정

- Graph 도메인 테스트 코드 추가
- Section 생성자 변경
- Path 가 StationResponse가 아닌 Station을 가지게 변경

choihwan2 commented:
말해주신 도메인객체의 테스트코드를 추가해봤습니다. 작성하면서 궁금증이 든 부분이 API를 확인하는 인수테스트??(FareAcceptanceTest) 부분에서도 Graph 도메인 객체가 제대로 만들어지는지 확인하는 것 같은데 이렇게 따로 도메인 객체를 테스트하는 코드를 만들어줄 필요가 있을까요??

kriery reviewed (approved):
안녕하세요 파블로, 피드백 반영 잘 봤습니다.
특이 사항 없이 이대로 머지하면 되겠네요 👍 
다음 미션도 화이팅입니다!

kriery commented:
> 말해주신 도메인객체의 테스트코드를 추가해봤습니다. 작성하면서 궁금증이 든 부분이 API를 확인하는 인수테스트??(FareAcceptanceTest) 부분에서도 Graph 도메인 객체가 제대로 만들어지는지 확인하는 것 같은데 이렇게 따로 도메인 객체를 테스트하는 코드를 만들어줄 필요가 있을까요??

인수검사는 **시스템 전체**를 놓고 의도한 대로 결과가 도출되는지를 테스트 하는 것인데요.
결과를 확인하는 것이라 정작 내부에서 단위 기능이 잘 동작하는지 어떤지는 알 수가 없어요.
단위 모듈에서 Test-case를 제공하는 것은 두 가지 합리적인 이유가 있습니다.
1. 모듈의 로직이 정상적으로 동작하는지 확인
2. 모듈의 스펙과 바운더리를 지속적으로 검증

인수 테스트에서는 이 내용이 [Controller - Service - Domain] 전체를 관통한 로직에 대한 검증이지만,
단위 테스트에서는 이 내용이 한 단위에 국한됩니다.

(이번에 작성해 주신) `GraphTest`에서
```Java
    @Test
    void 존재하지않는_경로_검사() {
        assertThatExceptionOfType(NotExistsDataException.class).isThrownBy(() -> 지하철노선그래프.getPathStations(증미역, 남부터미널역));
    }
```
이런 테스트는 "있을 수 없는 경로를 찾는" 입력에 대해서, `Graph` 객체가 Exception을 발생한다고 이해할 수 있습니다.
만약 이와 같은 테스트가 인수 검사에 있다고 하면, Exception을 발생하는 주체가 controller인지 service인지 domain인지 알 수는 없게 되죠.
(하지만 전체 시스템중 어디선가 확실히 발생한다고 이해할 수는 있습니다)

전체 시스템을 구성하고 나서 단위 테스트를 설계하면 자칫 단위 테스트가 필요 없다 여겨질 수 있긴 하지만.
각 모듈의 로직을 구상한 후, 실재로 구현하기 전에 단위 테스트로 바운더리를 정해놓고 개발하는 습관을 들여보면 어떨까요 (이게 참 저도 잘 안됩니다만)

choihwan2 commented:
그렇군요... 테스트에 대해서 좀 더 잘 이해하게 된것 같네요! 생각해보니 페어랑 테스트코드를 돌리면서도 '이게 어디에서 예외가 나온거지??' 라고 말한 경우가 많았는데, 이렇게 도메인 테스트를 따로 만들어놓게 되면 한눈에 확인 할 수 있겠네요. **단위테스트**의 중요성을 더 알고 가는것 같습니다! 감사합니다!
